include/swift/AST/Types.h

```diff
@@ -2196,33 +2196,39 @@ typedef ArrayRefView<Type,CanType,getAsCanType> CanTypeArrayRef;
 /// given type arguments.
 class BoundGenericType : public NominalOrBoundGenericNominalType,
     public llvm::FoldingSetNode {
-  
+
   /// Retrieve the intrusive pointer storage from the subtype
-  const Type *getTrailingObjectsPointer() const;
-  Type *getTrailingObjectsPointer() {
+  const SubstitutionMap *getTrailingObjectsPointer() const;
+  SubstitutionMap *getTrailingObjectsPointer() {
     const BoundGenericType *temp = this;
-    return const_cast<Type *>(temp->getTrailingObjectsPointer());
+    return const_cast<SubstitutionMap *>(temp->getTrailingObjectsPointer());
   }
 
 protected:
   BoundGenericType(TypeKind theKind, NominalTypeDecl *theDecl, Type parent,
-                   ArrayRef<Type> genericArgs, const ASTContext *context,
+                   SubstitutionMap Substitutions, const ASTContext *context,
                    RecursiveTypeProperties properties);
 
 public:
-  static BoundGenericType* get(NominalTypeDecl *TheDecl, Type Parent,
-                               ArrayRef<Type> GenericArgs);
+  static BoundGenericType *get(NominalTypeDecl *TheDecl, Type Parent,
+                               ArrayRef<Type> GenericArgs,
+                               ModuleDecl *M = nullptr);
+  static BoundGenericType *get(NominalTypeDecl *TheDecl, Type Parent,
+                               SubstitutionMap Substitutions);
 
-  /// Retrieve the set of generic arguments provided at this level.
-  ArrayRef<Type> getDirectGenericArgs() const {
-    return {getTrailingObjectsPointer(), Bits.BoundGenericType.GenericArgCount};
+  /// Retrieve the substitution map applied to the declaration's underlying
+  /// to produce the described type.
+  SubstitutionMap getSubstitutionMap() const {
+    return *getTrailingObjectsPointer();
```

Comment:
A SubstitutionMap is pointer-sized and always present; why does it need to be stored in the trailing objects?

Reply:
Good call!